### PR TITLE
Local test refactor, fixes #371, fixes #381

### LINKED ISSUE #371: Refactor local_test.go to make tests (more) idempotent
## What happened (or feature request):

As noted in https://github.com/drud/ddev/pull/337#issuecomment-312958420 there are some serious problems with the structure of local_test.go:

- TestMain sets up the workspaces and then removes them after tests. 
- However, TestMain does *not* shut down the sites or even try to do so. It could at least use the same techniques as TestLocalRemove and TestCleanupWithoutCompose to clean up the containers.
- The result is 3 nonworking sites (which do not show up on ddev list, apparently because the workspace is gone)
- TestLocalExec is dependent on a database having been loaded by TestLocalImportDb

## What you expected to happen:

* Each test should be able to be run independently
* Each test run should leave the system in a predictable state
* `ddev list` should report sites even when their workspace is trashed, this will likely happen in real life. 

## How to reproduce this:

`make testpkg TESTARGS="-run TestLocalStart"`

## Related source links or issues:

### LINKED ISSUE #381: ddev stop fails if a container is unhealthy
## What happened (or feature request):

I tried to stop a broken site (mounts broken underneath db container)

```
$ ddev stop
Failed to stop containers for drupal8: site does not appear to be running - web container db service unhealthy
```

Even though the site shows (with a good diagnosis) on `ddev list`, there's nothing I can do in ddev to solve this problem. I have to use the docker hammer.

## What you expected to happen:

I would think ddev would know how to stop a broken site. And that might even fix it! I note that `ddev rm` works as expected.

## Version: Please include the output of `ddev version` and the project's .ddev/config.yaml.

v0.8

## How to reproduce this:

`ddev start`
rm -rf ~/.ddev
ddev list

## Related source links or issues:

### PR DESCRIPTION
**Pull request by tannerjfco**
## The Problem:
OP #371. Our tests in the platform package are interdependent, requiring the full package tests to be run in order to run a given individual test.

## The Fix:
This PR breaks most of the test interdependencies. It moves start and remove functionality to the provision/cleanup handled in TestMain, so that individual tests can run without having to test the whole package. It also updates a couple tests, notably TestLocalExec, so that they can run without dependency on tests previous.

This also resolves #381, allowing ddev stop to run successfully unless there are no containers found.

There are a couple smaller tests I did not update to run independently, TestGetAppsEmpty and TestRouterNotRunning. These tests both validate a state in which no applications are running. I can update them if there's a strong desire, but I figured there's not much value in running these tests individually. 

## The Test:

Aside from TestGetAppsEmpty and TestRouterNotRunning, you should be able to pick an individual test out of local_test.go and run it. Sites should spin up for the test, it should pass, and then sites should tear down.

## Automation Overview:

This is largely an automation-specific improvement.

## Related Issue Link(s):
#371 OP for test refactor
#381 - ddev stop only works if site is in running state.

### CODE REVIEW

**Comment by rfay** [comment]
Well YEEHAW! Thanks!

File: pkg/plugins/platform/local_test.go [hunk 1]
```diff
@@ -62,13 +68,56 @@ func TestMain(m *testing.M) {
...
+			log.Fatal(err)
+		}
+
+		err = app.Down(true)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		runTime()
+		site.Cleanup()
```
Comment (rfay) [inline pkg/plugins/platform/local_test.go]: It sure is a huge win doing the container stops before the site cleanups! Thanks!
I'm not entirely sure why we don't consolidate the app-down functionality into the site.Cleanup(), but this works for me.

File: pkg/plugins/platform/local_test.go [hunk 2]
```diff
@@ -62,13 +68,56 @@ func TestMain(m *testing.M) {
...
+		err = app.Start()
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		runTime()
+		switchDir()
 	}
 
 	log.Debugln("Running tests.")
```
Comment (rfay) [inline pkg/plugins/platform/local_test.go]: Somewhere in here the sizing of the sites array that's done by TestLocalSetup() should be done if -short is specified.

Comment (beeradb) [inline pkg/plugins/platform/local_test.go]: Sadly, `-short` is not available in TestMain, or I would have done the sizing there in the first place :(

We might be able to parse go test args ourselves, but that feels really really ugly.

**Comment by rfay** [comment]
I'd love to see this land today so I can rebase it into Windows testing changes.

**Comment by rfay** [comment]
BTW, I think I'll be forced to implement test download (and extraction) caching just to speed up my windows work. Completely amazing how long tar extraction takes on Windows.

**Comment by tannerjfco** [comment]
This has been updated based on the feedback. "short mode" is set with an Env Var and handled in TestMain as opposed to the separate test function. Additional explicit cleanup of site data dirs added, as well as some other fixes/cleanup to the tests.

**Comment by rfay** [comment]
Going to give me lots of rebasing to do :)

**Pull request merged**